### PR TITLE
Bypass rollback

### PR DESCRIPTION
**Pull request by Vishtar**

### CODE REVIEW

File: src/services/bot/utils/getNotifications.ts [hunk 1]
```diff
@@ -27,6 +28,15 @@ export async function* getNotifications(
       const addressJettonsFromDb = await getJettonsFromDB(wallet.id)
       for (const jetton of addressJettonsFromDb) {
         if (!addressJettonsFromChainObj[jetton.token]) {
+          const firstPurchase = await getFirstAddressJettonPurchaseFromDB(jetton.id)
+          if (!firstPurchase) {
+            continue
+          }
+          const secondsFromPurchase = Date.now() / 1000 - firstPurchase.timestamp
+          // Estimated time to bypass a rollback
+          if (secondsFromPurchase <= 60) {
```
Comment (klntsky) [inline src/services/bot/utils/getNotifications.ts:37]: Please move to an env variable

File: src/services/bot/utils/getNotifications.ts [hunk 2]
```diff
@@ -27,6 +28,15 @@ export async function* getNotifications(
       const addressJettonsFromDb = await getJettonsFromDB(wallet.id)
       for (const jetton of addressJettonsFromDb) {
         if (!addressJettonsFromChainObj[jetton.token]) {
+          const firstPurchase = await getFirstAddressJettonPurchaseFromDB(jetton.id)
+          if (!firstPurchase) {
```
Comment (klntsky) [inline src/services/bot/utils/getNotifications.ts:32]: Is this possible? If we think it's not, we need to log this.

File: src/utils/parseTxData/api.ts [hunk 3]
```diff
@@ -21,7 +21,9 @@ export const api = async (address: string) => {
     res[jettonInfo.symbol] = {
       ...jettonInfo,
       pnlPercentage,
-      chart: (slicedChart.length >= 2 ? slicedChart : chart).reverse().map(entity => [entity[0], normalizePrice(entity[1], jettonInfo.decimals)]),
+      chart: (slicedChart.length >= 2 ? slicedChart : chart)
```
Comment (klntsky) [inline src/utils/parseTxData/api.ts:24]: If `slicedChart` does not have at least two elements, that means our algorithm is wrong. log it as error.

File: src/utils/parseTxData/api.ts [hunk 4]
```diff
@@ -21,7 +21,9 @@ export const api = async (address: string) => {
...
+        .reverse()
+        .map(entity => [entity[0], normalizePrice(entity[1], jettonInfo.decimals)]),
```
Comment (klntsky) [inline src/utils/parseTxData/api.ts:26]: This logic should not be here. return the literal price value, and normalize it on the frontend. Less load on the server + eventually I want to turn it into a string (not valid number at all), e.g. something like `0.₉1234`

**Review by klntsky — CHANGES_REQUESTED** [review]

File: src/utils/parseTxData/api.ts (hunk 3 above)
Comment (klntsky) [reply src/utils/parseTxData/api.ts:24]: Do not reverse the chart here. Reverse it on the frontend

File: src/utils/parseTxData/api.ts (hunk 4 above)
Comment (Vishtar) [reply src/utils/parseTxData/api.ts]: Actually, frontend is completely ok with displaying a string. But I guess our clients have decent hardware and don't care about some client-side transformations anyway.

File: src/utils/parseTxData/api.ts (hunk 3 above)
Comment (Vishtar) [reply src/utils/parseTxData/api.ts]: I just didn't touch that, but sure.

File: src/services/bot/utils/getNotifications.ts (hunk 2 above)
Comment (Vishtar) [reply src/services/bot/utils/getNotifications.ts]: This is technically possible. if we have a race condition.
I replaced multiple queries with a single postgres transaction with these queries to completly avoid that possibility now.

**Pull request merged**